Web console: editing compute resource limits

spadgett opened the pull request:
* Allow editing compute resource requests and limits for deployment configs and replication controllers that don't have a deployment config.
* Allow setting compute resource requests and limits during create.

https://trello.com/c/23NdtCwN

- [x] Add compute resources to create form
- [x] Disable form on submit
- [x] Update application generator unit tests
- [x] Update for #6901
- [x] Don't require limit if a default is set
- [x] https://bugzilla.redhat.com/show_bug.cgi?id=1303547

Browse deployment config:

<img width="757" alt="openshift_web_console" src="https://cloud.githubusercontent.com/assets/1167259/12623359/c6770294-c4f7-11e5-833e-06802e651adf.png">

Set resource limits page:

<img width="844" alt="openshift_web_console" src="https://cloud.githubusercontent.com/assets/1167259/12623383/e8f4b85c-c4f7-11e5-84f6-19c7becd63a5.png">

/cc @jwforres @derekwaynecarr @sosiouxme

spadgett commented:
@jwforres PTAL

jwforres commented on assets/app/scripts/controllers/create/createFromImage.js:
```diff
@@ -99,6 +102,12 @@ angular.module("openshiftConsole")
         }
 
+        var watches = [];
+        watches.push(DataService.watch("limitranges", context, function(limitRanges) {
```
limit ranges very very rarely change, wondering if we should just list here and save a websocket connection,  when we originally talked about watching for this stuff it was for quota which was going to be updating potentially.

spadgett replied:
Yeah, I'll make the change.

spadgett commented:
@jwforres thoughts on this alternative for showing limit range and default?

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/12659004/5c85cce2-c5da-11e5-88bb-532661b02a50.png)

spadgett commented:
Hm, maybe the placeholder is a bad idea for the default.

jwforres commented on assets/app/scripts/controllers/setLimits.js:
```diff
@@ -0,0 +1,100 @@
+        );
+
+        var watches = [];
+        watches.push(DataService.watch("limitranges", context, function(limitRanges) {
```
same thing here, change to a list

spadgett commented:
Rebased and removed the limit range watches.

jwforres commented on assets/app/scripts/directives/editRequestLimit.js:
```diff
@@ -0,0 +1,194 @@
+        var validateLimitRange = function() {
+          var value, minValid = true, maxValid = true;
+          if (scope.amount) {
+            value = usageValue(scope.amount + scope.unit, scope.unit);
```
trying to follow exactly where unit normalization is happening with the current value vs the min and max.  some comments might help here

spadgett commented:
Rebased and updated. Current UI:

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/12677260/acc8190e-c665-11e5-954c-c1a0c5c34ea8.png)

@jwforres please have another look

spadgett commented:
[test]

jwforres commented on assets/app/scripts/directives/editRequestLimit.js:
```diff
@@ -0,0 +1,234 @@
+        }
+
+        var validateLimitRange = function() {
+          // Use usageValue filter to normalize units for comparison.
```
thanks this section is a lot easier to follow now

jwforres commented on assets/app/scripts/directives/editRequestLimit.js:
```diff
@@ -0,0 +1,234 @@
+
+            // Limit must be within the max limit/request ratio if defined.
+            if (scope.maxLimitRequestRatio) {
+              limitWithinRatio = (limit / request) <= scope.maxLimitRequestRatio;
```
when request is set and limit isnt then you have an unbounded limit, if a a maxLimitRequestRatio is set then this pod would be rejected ( i just double checked with @derekwaynecarr )

spadgett replied:
Fixed.

spadgett replied:
@jwforres It looks like limit is not required when there's a default, even with `maxLimitRequestRatio` set.

jwforres replied:
Yeah it is only an issue when there is no default
On Feb 2, 2016 1:27 PM, "Sam Padgett" <notifications@github.com> wrote:

> In assets/app/scripts/directives/editRequestLimit.js
> <https://github.com/openshift/origin/pull/6853#discussion_r51611306>:
>
> > +        };
> > +
> > +        var validateLimitAgainstRequest = function() {
> > +          // Use usageValue filter to normalize units for comparison.
> > +          var limit = scope.amount && usageValue(scope.amount + scope.unit),
> > +              request = scope.request && usageValue(scope.request),
> > +              limitLargerThanRequest = true,
> > +              limitWithinRatio = true;
> > +
> > +          if (request && limit) {
> > +            // Limit must be greater than or equal to request.
> > +            limitLargerThanRequest = limit >= request;
> > +
> > +            // Limit must be within the max limit/request ratio if defined.
> > +            if (scope.maxLimitRequestRatio) {
> > +              limitWithinRatio = (limit / request) <= scope.maxLimitRequestRatio;
>
> @jwforres <https://github.com/jwforres> It looks like limit is not
> required when there's a default, even with maxLimitRequestRatio set.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6853/files#r51611306>.
>

spadgett replied:
I'll fix.

jwforres commented on assets/app/scripts/directives/editRequestLimit.js:
```diff
@@ -0,0 +1,234 @@
+          scope.limits = {};
+          angular.forEach(scope.limitRanges, function(limitRange) {
+            angular.forEach(limitRange.spec.limits, function(limit) {
+              // Find the largest minimum.
```
should this be in a LimitRanges service, LimitRanges.getEffectiveLimits() or something.  Feels like we are going to need this somewhere else.

spadgett replied:
Added a service.

jwforres commented on assets/app/scripts/directives/editRequestLimit.js:
```diff
@@ -0,0 +1,234 @@
+          angular.forEach(scope.limitRanges, function(limitRange) {
+            angular.forEach(limitRange.spec.limits, function(limit) {
+              // Find the largest minimum.
+              if (limit.min && isLarger(limit.min[scope.type], scope.cpuMin)) {
```
scope.min instead of scope.cpuMin?  since it might be the memory min right?

jwforres replied:
yeah this seems to be a typo, you are setting scope.min below

spadgett replied:
Yeah, it's a typo. I'll fix.

jwforres commented on assets/app/scripts/directives/editRequestLimit.js:
```diff
@@ -0,0 +1,234 @@
+                scope.max = limit.max[scope.type];
+              }
+              if (limit['default']) {
+                scope.defaultLimit = limit['default'][scope.type] || scope.defaultLimit;
```
add a TODO here that we need to revisit this when we decide what the correct indeterminate behavior is for multiple valid defaults

spadgett replied:
Added the TODO comment.

jwforres commented on assets/app/scripts/filters/util.js:
```diff
@@ -105,6 +105,12 @@ angular.module('openshiftConsole')
+
+      // No normalization of unit.
+      if (!humanizeUnits) {
+        return split.slice(1);
```
So this is stripping the unit and only returning the value?  I think I would rather have usageValue take a flag that prevents normalization.  It's weird that the default behavior of a function called amountAndUnit is to drop the unit.

spadgett replied:
> So this is stripping the unit and only returning the value?

No, it's returning the amount (index 1) and unit (index 2).

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/12680436/cabbce62-c677-11e5-9f66-6a11f9881e05.png)

spadgett replied:
I'll add a comment.

jwforres replied:
ah duh ok

spadgett replied:
It's confusing. I had to set a breakpoint to remember why I needed the slice.

spadgett replied:
I've updated the code so things are more clear.

jwforres commented on assets/app/styles/_core.less:
```diff
@@ -517,6 +517,15 @@ label.checkbox {
 
+.compute-resource .inline-select {
+  // Spacing below input at mobile sizes.
+  margin-top: 5px;
```
Looks like this is always added, not just at mobile, fix comment?

spadgett replied:
No, it's reset to 0 just below at screen-sm-min.

spadgett commented:
@jwforres comments addressed

jwforres commented:
LGTM fork_ami

spadgett commented:
Rebased and fixed java/bindata conflict.

spadgett commented:
Rebased.

sosiouxme commented:
@spadgett FYI I think the PodLimitRequest mechanics are firmed up, will find out more solidly at scrum this afternoon.

See implementation details under https://trello.com/c/rRN7m0zv

PR https://github.com/openshift/origin/pull/6901

spadgett commented:
Rebased.

spadgett commented:
[test]

jwforres commented on assets/app/scripts/services/limits.js:
```diff
@@ -0,0 +1,189 @@
+      angular.forEach(containers, function(container) {
+        var resources = container.resources;
+        if (!resources) {
+          return;
```
wouldn't this container's defaults still matter, you wouldn't have validation on these fields but the defaults should still go against the total

jwforres commented:
LGTM [merge]

liggitt commented on assets/app/scripts/controllers/setLimits.js:
```diff
@@ -0,0 +1,107 @@
+      $scope.resourceURL = Navigate.resourceURL($scope.name, "ReplicationController", $routeParams.project);
+      $scope.showPodWarning = true;
+    } else {
+      Navigate.toErrorPage("A replication controller or deployment config must be provided.");
```
return here?

liggitt commented on assets/app/scripts/controllers/setLimits.js:
```diff
@@ -0,0 +1,107 @@
+
+    $scope.hideCPU = LimitRangesService.hideCPUComputeResources();
+
+    var type, displayName;
```
nit, resource, not type
